src/Compiler/NodeVisitor/AddFormClassesNodeVisitor.php

```diff
@@ -0,0 +1,181 @@
+            }
+
+            if (!$node->var instanceof FuncCall) {
+                return null;
+            }
+
+            if ($this->nameResolver->resolve($node->var) !== 'end') {
+                return null;
+            }
+
+            $fieldName = $node->dim->value;
+            $formField = $this->actualForm->getFormField($fieldName);
+            if ($formField === null) {
+                $rootParentNode = $node;
+                while (true) {
+                    $parentNode = $rootParentNode->getAttribute('parent');
+                    if ($parentNode === null) {
+                        break;
+                    }
+
+                    $rootParentNode = $parentNode;
+
+                    if ($rootParentNode instanceof Echo_ || $rootParentNode instanceof If_) {
+                        break;
+                    }
+                }
+                $this->errorFieldNodes[] = [
+                    'node' => $rootParentNode,
+                    'field' => $fieldName,
+                ];
+                return null;
+            }
+
+            $node->var = new Variable('form');
+            return $node;
+        }
+
+        return null;
+    }
+
+    public function leaveNode(Node $node)
+    {
+        foreach ($this->errorFieldNodes as $errorFieldNode) {
+            if ($errorFieldNode['node'] === $node) {
+                $error = new Error('Form field with name "' . $errorFieldNode['field'] . '" probably does not exist.');
+                $errorNode = $error->toNode();
+                $errorNode->setAttributes($node->getAttributes());
+                return $errorNode;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param Node[] $nodes
+     * @return Node[]
+     */
+    public function afterTraverse(array $nodes): array
+    {
+        $componentType = '\Nette\ComponentModel\IComponent';
+        foreach ($this->forms as $formName => $form) {
+            $className = $this->formClassNames[$formName] ?? null;
+            if ($className === null) {
+                continue;
+            }
+            $method = (new Method('offsetGet'))
+                ->addParam(new Param('name'))
+                ->addStmts([
+                    new Return_(
+                        new StaticCall(
+                            new Name('parent'),
+                            new Identifier('offsetGet'),
+                            [
+                                new Arg(new Variable('name')),
+                            ]
+                        )
+                    ),
+                ])
+                ->makePublic()
+                ->setReturnType($componentType);
+            $comment = '@return ' . $componentType;
+            foreach ($form->getFormFields() as $formField) {
+                $comment = str_replace($componentType, '($name is \'' . $formField->getName() . '\' ? ' . $formField->getTypeAsString() . ' : ' . $componentType . ')', $comment);
```

Comment:
Maybe it would be better to use some placeholder like `%%TYPE%%` here instead of `$componentType` - theoretically `$componentType` could appear as part of namespace.

Reply:
I have to check what it is :) this part is just copy-pasted from old code